### PR TITLE
migrate types from transaction/{core->types}

### PR DESCRIPTION
**Pull request by ryankurte**
migrating transaction related types / constants / etc. from `mc-transaction-core` to `mc-transaction-types` where possible.

this is the first part of refactoring `TxSummary` computation into a separate `mc-transaction-summary` crate to simplify compilation for hardware wallets, and another step in the upstreaming of [ryan/ledger-wip](https://github.com/mobilecoinfoundation/mobilecoin/compare/ryan/ledger-wip).

blocked on:
- [x] #3083

### CODE REVIEW

**Review by cbeck88 — CHANGES_REQUESTED** [review]
Requested to back out of the "Summarize" trait

File: transaction/core/src/tx_summary.rs [hunk 1]
```diff
@@ -86,50 +86,26 @@
...
-    #[prost(uint64, tag = "4")]
-    pub fee_token_id: u64,
+pub use mc_transaction_types::tx_summary::*;
 
-    /// The block index at which this transaction is no longer valid.
-    #[prost(uint64, tag = "5")]
-    pub tombstone_block: u64,
+/// [Summarize] trait for constructing [TxSummary] object from foreign types
+pub trait Summarize {
+    /// Make a TxSummary for a given TxPrefix and pseudo-output commitments
```
Comment (cbeck88) [inline transaction/core/src/tx_summary.rs:101]: It looks to me that we are taking the `TxSummary::new` function and reorganizing it as a trait, implemented on a pair of references.

I can't see any reasons why this is better and I see a lot of reasons why it's worse. Can we back out if this change?

Comment (cbeck88) [reply transaction/core/src/tx_summary.rs:101]: I see, to break dependencies?

Comment (cbeck88) [reply transaction/core/src/tx_summary.rs:101]: Okay, can we maybe do this like

`impl TryFrom<(&...,&...)> for TxSummary`?

I think that's more idiomatic and its more helpful to someone reading rustdoc documentation

Comment (ryankurte) [reply transaction/core/src/tx_summary.rs:101]: so i need to move the `TxSummary` objects to types to share em between transaction/{core/extra/summary} but the actual logic should stay in core (to be moved to summary), so the trait is a trick to allow impl on a foreign type...

open to alternatives? or i could refactor the `Summarize` trait to act on the `TxSummary` object?

```rust
pub trait Summarize {
  fn new(&TxPrefix, &[CompressedCommitment]) -> Result<TxSummary, ZipExactError> { .. }
}
```

Comment (cbeck88) [reply transaction/core/src/tx_summary.rs:101]: Right -- what I said doesnt work with orphan rules in the way you're shooting for.

What if we spelled the trait `TxSummaryNew` similar to Tony arcieri's `AeadNew` trait? I feel like calling it `Summarize` isn't very clear. Lots of things can be "summarized" but this is really specifically meant to be the constructor for TxSummary.

Comment (ryankurte) [reply transaction/core/src/tx_summary.rs:101]: yeah good call, swapped to `TxSummaryNew` ^_^

**Review by cbeck88 — COMMENTED** [review]
looks good to me so far! thank you

**Review by cbeck88 — APPROVED** [review]

**Comment by briancorbin** [comment]
Current dependencies on/for this PR:
* master
  * **PR #3083** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/3083" target="_blank">Graphite</a> 
    * **PR #3125** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/3125" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/3125?utm_source=stack-comment).

**Review by dorebell — APPROVED** [review]
Looks good to me!

**Review by briancorbin — APPROVED** [review]

**Review by jcape — COMMENTED** [review]
Could you update this to follow the style guide: https://github.com/mobilecoinfoundation/mobilecoin/blob/master/CONTRIBUTING.md#coding-style

File: transaction/types/src/lib.rs [hunk 2]
```diff
@@ -1,18 +1,34 @@
...
+extern crate alloc;
+
+pub use amount::{Amount, AmountError};
+pub use block_version::{BlockVersion, BlockVersionError, BlockVersionIterator};
+#[cfg(feature = "alloc")]
+pub use masked_amount::{MaskedAmount, MaskedAmountV1, MaskedAmountV2};
+pub use token::TokenId;
+#[cfg(feature = "alloc")]
+pub use tx_summary::{TxInSummary, TxOutSummary, TxSummary};
+pub use unmasked_amount::UnmaskedAmount;
```
Comment (jcape) [inline transaction/types/src/lib.rs:20]:
```suggestion
pub use crate::{
    amount::{Amount, AmountError},
    block_version::{BlockVersion, BlockVersionError, BlockVersionIterator},
    unmasked_amount::UnmaskedAmount,
    token::TokenId,
};
#[cfg(feature = "alloc")]
pub use crate::{
    masked_amount::{MaskedAmount, MaskedAmountV1, MaskedAmountV2},
    tx_summary::{TxInSummary, TxOutSummary, TxSummary},
};
```

File: transaction/types/src/unmasked_amount.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Unmasked amount types
+use crate::{amount::Amount, TokenId};
+
```
Comment (jcape) [inline transaction/types/src/unmasked_amount.rs:5]:
```suggestion
```

**Review by jcape — APPROVED** [review]
LGTM, but some of the alloc-vs-no_alloc exports could be made more clear...

**Pull request merged**